taichi/backends/metal/codegen_metal.cpp

```diff
@@ -449,15 +470,50 @@ class KernelCodegenImpl : public IRVisitor {
       const auto *argload = stmt->base_ptrs[0]->as<ArgLoadStmt>();
       const int arg_id = argload->arg_id;
       const int num_indices = stmt->indices.size();
-      std::vector<std::string> size_var_names;
-      for (int i = 0; i < num_indices; i++) {
-        std::string var_name = fmt::format("{}_size{}_", stmt->raw_name(), i);
+      const auto &element_shape = stmt->element_shape;
+      std::vector<std::string> size_exprs;
+      enum ExternalArrayLayout { layout_AOS = 0, layout_SOA = 1 };
+      const auto layout = stmt->element_dim <= 0 ? layout_AOS : layout_SOA;
+
+      // Args buffer arrange dimensions from outer to inner
+      // AoS args buffer:   array_shape|element_shape
+      // SoA args buffer: element_shape|array_shape
+      //
+      // ti.Matrix.ndarray(3, 2, ti.f32, (5, 4), layout=ti.Layout.AOS)
+      // args buffer: 5, 4, 3, 2
+      // ti.Matrix.ndarray(3, 2, ti.f32, (5, 4), layout=ti.Layout.SOA)
+      // args buffer: 3, 2, 5, 4
+      const int arr_shape_len = num_indices - element_shape.size();
+      int index_i = 0;
+      const auto add_elem_shape_exprs = [&]() {
+        for (int es : element_shape) {
+          size_exprs.push_back(std::to_string(es));
+          ++index_i;
+        }
+      };
+      int arr_shape_offset = 0;
+      if (layout == layout_SOA) {
+        add_elem_shape_exprs();
+        // When the layout is SOA, element shape comes before array shape, so
+        // we have to skip the element shapes first.
+        // TODO: Element shape is a compile-time known information, so extra
+        // args will always only need the array shape.
+        arr_shape_offset = element_shape.size();
```

Comment:
@ailzhang @jim19930609 @strongoier  We should consider removing element_shape from extra_args...

Reply:
agreed! We might hit some unexpected usage of element_shape in extra_args along the cleanup (probably need to check spirv codegen first to see ndarray was implemented there.